lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -416,3 +447,100 @@ def get_num_raft_peers(self) -> int:
         """Return the number of raft peers."""
         raft_config = self._client.sys.read_raft_config()
         return len(raft_config["data"]["config"]["servers"])
+
+    def _get_autounseal_policy_name(self, relation_id: int) -> str:
+        """Return the policy name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _get_autounseal_approle_name(self, relation_id: int) -> str:
+        """Return the approle name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _configure_autounseal_policy(self, mount: str, relation_id: int, key_name: str) -> str:
```

Comment:
I'm not against the latter suggestion, but I dislike the first. It doesn't scale. We shouldn't have a function called `_configure_policy` that just grows to contain an ever-expanding list of kwargs that may or may not be relevant to the policy being configured. It doesn't make a nice interface. `mount` is already specific to KV, and if we include the autounseal policy it will also need `key_name`. That doesn't seem right to me. 

It also doesn't allow for things to be inferred without additional input from the caller (such as the policy name, in this case). 

I guess it comes down to how much logic we want to push to the caller. We can remove the kwargs entirely, but then the caller has to fill in the policy parameters themselves. We can not have an autounseal-specific policy function, but then we push this logic to the caller.

I thought the goal of this library was not necessarily just to be an interface to vault, but to avoid repeating logic between the k8s and machine charms. The more we push to the callers, the more work we have to repeat in multiple places.

I'm curious what your thoughts are on this perspective.

Reply:
My perspective is that a `policy` is a `policy` no matter what utilizes it. `mount` for example isn't KV specific in general, we can have a custom mount for any policy we want to template. I would agree with you if `_configure_policy` is just a massive switch statement or configuring policies for each thing looked completely different, but in cases where we have 3 different functions and they are just copy and pasted except only 1 line or 1 argument is different, I feel like that's antithetical to what a function is. That is exactly when we should be refactoring multiple functions into one.

However, if we prefer the comfort of that over having a succinct and correct function, let's go with it.

When it comes to sharing code between machine and k8s charms, I don't think this is directly relevant since both charms will use the exact same options whatever we choose. However, in the event where we will need unique functions for either, I think it doesn't make sense to dump specific functions into shared libraries unless those functions are going to be shared verbatim. If we have `func_machine` and `func_k8s` instead of `func(options)`, then there is no point in sharing these functions anyway.

Reply:
I spent some time looking at this and I still believe our current implementation of `configure_policy` isn't the best direction forward, but I think I'll do my best to conform for now and suggest an improvement afterwards.

I think that that a refactor is probably too big to include in this PR, and I don't want to fragment the code with two different ways of doing things.

Reply:
I had done the PR to move to that [here](https://github.com/canonical/vault-k8s-operator/pull/234/files), which received no comments there. If we as a team decide otherwise, I have no problem switching back.

Reply:
It's not about "switching back", it's about an alternative way to move forward. There's nothing wrong with the way it was implemented at the time, it just makes some assumptions that make it hard to re-use in new use cases.

I tried to make the code conform last night and although `configure_policy()` was a little unfortunate for the reasons mentioned above, `configure_approle()` gets messy too because we've made some assumptions about the policy as well. For example, we don't want to set a `token_ttl`, or `token_max_ttl`. We want to set `token_period` instead. Can we start adding some branching to `configure_approle()` based on our business logic? Absolutely. But that gets really difficult to follow or keep correct as this continues to absorb new use cases.

Reply:
Deferring this discussion to a spec: https://warthogs.atlassian.net/browse/TLSENG-269

Reply:
> I tried to make the code conform last night and although `configure_policy()` was a little unfortunate for the reasons mentioned above, `configure_approle()` gets messy too because we've made some assumptions about the policy as well. For example, we don't want to set a `token_ttl`, or `token_max_ttl`. We want to set `token_period` instead. Can we start adding some branching to `configure_approle()` based on our business logic? Absolutely. But that gets really difficult to follow or keep correct as this continues to absorb new use cases.

It's more about `token_ttl: Optional = None`. Branching statements that increase complexity and separating functions by copy pasting most of it and changing a couple of lines is a false dichotomy. There are many ways to make functions more dynamic without branching if statements, and many ways to split the same functionality into two different most common use-case functions with minimal arguments. 

In this case I believe it's preferable to share the function. If we need to add an extra argument so you can still have the function signature you want, but we need to add 2 `"60s"`'s into other invocations, that's great. In fact adding 60s's to other invocations is better than having 60s as default, it's more explicit so in my eyes that's a win-win situation. If we make two functions that have different arguments to the same hvac call, that's just 1 win. And in this case, I believe that two functions share more similarities than not.

Feel free to resolve this when you read it, I agree that a spec is probably a better place to talk about this, and there's really no need to hold this PR up.

Reply:
I'm not suggesting copy/pasting or repeating parts anywhere. I could have factored out the common parts for sure, and that's direction I would like to move. But, as they're written, they're not easy to re-use. If we continue to add more variables every time we want different values, this library effectively becomes a thin wrapper around hvac, with the exact same parameters, just surrounded by a try..except block.

Keep in mind, also, that a majority of this was written before your refactor as I started to work on this back in February before I had to put it aside for Manual Unseal. My bad for not updating it in consideration of the new changes, but hopefully that provides some context.

Reply:
> this library effectively becomes a thin wrapper around hvac, with the exact same parameters, just surrounded by a try..except block.

you say that like it's a bad thing :D I think a bunch of my comments really stems from the fact that I do want this. and I only commented instead of asking for changes because I'm aware that this PR is not the right place to tackle this. In the future I think I'm going to avoid these types of comments and open a ticket for us to talk about it in the backlog grooming meeting.

Reply:
It was great conversation, so I appreciate it nonetheless.

Reply:
always love talking with you

Reply:
Fixed in https://github.com/canonical/vault-k8s-operator/pull/365/commits/e9f318763e6f85ed946049a862f15f5bb94885f2